src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPath.cs

```diff
@@ -142,23 +130,23 @@ public void WalkWith(PathSegmentHandler handler)
         internal bool Equals(ODataPath other)
         {
             ExceptionUtils.CheckArgumentNotNull(other, "other");
-            if (this.segments.Count != other.segments.Count)
+            if (this.Count != other.Count)
             {
                 return false;
             }
 
-            return !this.segments.Where((t, i) => !t.Equals(other.segments[i])).Any();
+            return !this.Where((t, i) => !t.Equals(other[i])).Any();
         }
 
         /// <summary>
         /// Add a segment to this path.
         /// </summary>
         /// <param name="newSegment">the segment to add</param>
         /// <exception cref="System.ArgumentNullException">Throws if the input newSegment is null.</exception>
-        internal void Add(ODataPathSegment newSegment)
+        internal new void Add(ODataPathSegment newSegment)

```

Comment:
@marabooy Why did you remove `new` modifier from here? Given that its used to instruct the compiler to use the child class implementation, you could easily change behaviour altogether

Reply:
reverted to using a backing list.